### ISSUE TITLE
Not calling `setGlobalDispatcher newDispatcher()` costs 500ms

### ISSUE THREAD

**Issue by RedBeard0531** [nim-lang/Nim#6846]
```nim
import asyncdispatch
import asyncfile

var asyncStdout = 1.AsyncFD.newAsyncFile()
proc doStuff: Future[void] {.async.} =
  await asyncStdout.write "hello world\n"

#setGlobalDispatcher newDispatcher() # magical Go Fast button!
waitFor doStuff()
```

time: 502ms
time with dispatcher: 2ms

**Comment by loongw** [comment]
The iterator generated by async like this:
```nim
  iterator doStuffIter245003(): FutureBase {.closure.} =
    var future245004 = asyncStdout.write "hello world\x0A"
    yield future245004
    future245004.read
    complete(retFuture245002)
```
But the `future245004` completed immediately after `asyncStdout.write`.
Then the next yield will cause an extra poll, which will get nothing but 500ms timeout.
I guess the right code should looks like this:
```nim
  iterator doStuffIter245003(): FutureBase {.closure.} =
    var future245004 = asyncStdout.write "hello world\x0A"
    if future245004.finished:
      complete(retFuture245002)
      return
    yield future245004
    future245004.read
    complete(retFuture245002)
```

**Comment by yglukhov** [comment]
Will be fixed by #6614

**Comment by dom96** [comment]
This appears to be fixed now.

Reproduces in 0.17.2 but not 0.18.0 nor ``devel``.
